test/boost/stall_free_test.cc

```diff
@@ -53,3 +55,523 @@ SEASTAR_THREAD_TEST_CASE(test_merge4) {
+            on_clear(v);
+            return make_ready_future<>();
+        }
+    };
+    int cleared_gently = 0;
+    std::unique_ptr<X> p = std::make_unique<X>(0, [&cleared_gently] (int) {
+        cleared_gently++;
+    });
+
+    utils::clear_gently(p).get();
+    BOOST_CHECK(p);
+    BOOST_REQUIRE_EQUAL(cleared_gently, 1);
+}
+
+SEASTAR_THREAD_TEST_CASE(test_clear_gently_shared_ptr) {
+    struct X {
+        int v;
+        std::function<void(int)> on_clear;
+        X(int i, std::function<void (int)> f) : v(i), on_clear(std::move(f)) {}
+        future<> clear_gently() {
+            on_clear(v);
+            return make_ready_future<>();
+        }
+    };
+    int cleared_gently = 0;
+    lw_shared_ptr<X> p0 = make_lw_shared<X>(0, [&cleared_gently] (int) {
+        cleared_gently++;
+    });
+
+    utils::clear_gently(p0).get();
+    BOOST_CHECK(p0);
+    BOOST_REQUIRE_EQUAL(cleared_gently, 1);
+
+    lw_shared_ptr<X> p1 = p0;
+
+    utils::clear_gently(p0).get();
+    BOOST_CHECK(p0);
+    BOOST_REQUIRE_EQUAL(cleared_gently, 1);
+    utils::clear_gently(p1).get();
+    BOOST_CHECK(p1);
+    BOOST_REQUIRE_EQUAL(cleared_gently, 1);
+}
+
+SEASTAR_THREAD_TEST_CASE(test_clear_gently_trivial_array) {
+    std::array<int, 3> a = {0, 1, 2};
+
+    utils::clear_gently(a).get();
+}
+
+SEASTAR_THREAD_TEST_CASE(test_clear_gently_non_trivial_array) {
+    struct X {
+        int v;
+        std::function<void(int)> on_clear;
+        X(int i, std::function<void (int)> f) : v(i), on_clear(std::move(f)) {}
+        future<> clear_gently() {
+            on_clear(v);
+            return make_ready_future<>();
+        }
+    };
+    constexpr int count = 3;
+    std::array<std::unique_ptr<X>, count> a;
+    int cleared_gently = 0;
+
+    for (int i = 0; i < count; i++) {
+        a[i] = std::make_unique<X>(i, [&cleared_gently] (int) {
+            cleared_gently++;
+        });
+    }
+
+    utils::clear_gently(a).get();
+    BOOST_REQUIRE_EQUAL(cleared_gently, count);
+}
+
+SEASTAR_THREAD_TEST_CASE(test_clear_gently_trivial_vector) {
+    constexpr int count = 100;
+    std::vector<int> v;
+
+    v.reserve(count);
+    for (int i = 0; i < count; i++) {
+        v.emplace_back(i);
+    }
+
+    utils::clear_gently(v).get();
+    BOOST_CHECK(v.empty());
+}
+
+SEASTAR_THREAD_TEST_CASE(test_clear_gently_trivial_small_vector) {
+    utils::small_vector<int, 1> v;
+    constexpr int count = 10;
+
+    v.reserve(count);
+    for (int i = 0; i < count; i++) {
+        v.emplace_back(i);
+    }
+
+    utils::clear_gently(v).get();
+    BOOST_CHECK(v.empty());
+}
+
+SEASTAR_THREAD_TEST_CASE(test_clear_gently_vector) {
+    static thread_local std::vector<int> res;
+    struct X {
+        int v;
+        X(int i) : v(i) {}
+        X(X&& x) noexcept : v(x.v) {}
+        future<> clear_gently() noexcept {
+            res.push_back(v);
+            return make_ready_future<>();
+        }
+    };
+    static_assert(std::is_trivially_destructible_v<X>);
```

Comment:
I want to verify that the big pile of template constraints in stall_free.hh actually works
and it doesn't recognize the vector as a `TriviallyClearableSequence` although X *is* trivially destructible.
Because `X` has a `clear_gently` method.
I'll add a comment about that.